### DISCUSSION TITLE
[Share] Copy Open Tabs to Clipboard as Markdown or HTML list

### DISCUSSION THREAD

**Discussion by s1291** [windingwind/zotero-actions-tags#206] [Action Scripts]
### Description

![image](https://github.com/windingwind/zotero-actions-tags/assets/17838489/82bf73e8-c66f-452e-9d8a-9c4c122eb3a0)

You can then paste the result to Zotero notes, obsidian, or any Markdown editor:

![image](https://github.com/windingwind/zotero-actions-tags/assets/17838489/54702cf0-ca61-4349-aff5-0a4b7980dbeb)

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Copy the open tabs to Clipboard as a Markdown or HTML list; ready to paste in Zotero notes or Obisidian.
 * @author Samir Ouchene, Thanks to FeralFlora for their hints.
 * @usage plug and play 
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/206
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/206
 */

const Zotero_Tabs = require("Zotero_Tabs");
//const window = require("window");
const clipboard = new Zotero.ActionsTags.api.utils.ClipboardHelper();

/** @type {'md' | 'html'} Change the type of the URL. */
URLType = 'html';

// For efficiency, only execute once for all selected items
if (item) return;
// Filter the tabs to include only the one with type 'reader' or 'reader-unloaded'
let tabs = Zotero_Tabs._tabs.filter(tab => 
['reader', 'reader-unloaded'].includes(tab.type)
);
let tabData_ = JSON.stringify(tabs);
tabData = JSON.parse(tabData_);
itemIDs = tabData.map(obj => obj.data.itemID);
itemTitles = tabData.map(obj => obj.title);
	
// get the item keys using the itemIDs
itemKeys = itemIDs.map(itemID => Zotero.Items['_objectCache'][itemID]['key']);
URIs = itemKeys.map(itemKey => `zotero://open-pdf/library/items/${itemKey}`);


if (URLType.toLowerCase() === 'md') {
    mimeStr = "text/unicode";
    listOfURIs = itemTitles.map((title,index) => `* [${title}](${URIs[index]})`).join('\n');
} else if (URLType.toLowerCase() === 'html') {
    mimeStr = "text/html";
    listOfURIs_ = itemTitles.map((title,index) => `<li><a href="${URIs[index]}"> ${title}</a></li>`).join('');
    listOfURIs = `<ul>${listOfURIs_}</ul>`;
}

clipboard.addText(listOfURIs, mimeStr);
clipboard.copy();
//window.alert("Successfully copied " + itemIDs.length + " tabs to clipboard");
return `[Copy Open Tabs] copied ${itemIDs.length} items to clipboard`;
```

**Comment by windingwind** [comment]
We'd better add `if (item) return;` at the beginning of the script to prevent it being executed for multiple times when multiple items are selected in the library.

> When selecting multiple items in the library, the action will be triggered for each item. You can use the items variable to get the selected items array and avoid duplicate operations.

See https://github.com/windingwind/zotero-actions-tags#custom-script > Items

**Reply by s1291** [reply]
Updated the script and added `if (item) return;`

Thank you

**Comment by mjthoraval** [comment]
Thank you very much for the script!
I understand that the `${title}` corresponds to the title of the tab at the moment.

Would it be possible to customize the format of the output with the QuickCopy formatted citations?
I would like to get something like:
[Quick Copy formatted citation] [Select link] [Open-pdf link]
- with no link on the [Quick Copy formatted citation] (see https://github.com/windingwind/zotero-actions-tags/discussions/165)
- with a select link (zotero://select) on the [Select link]
- with a open-pdf link (zotero://open-pdf) on the [Open-pdf link]
(see https://github.com/windingwind/zotero-actions-tags/discussions/115)
